packages/snaps-controllers/src/snaps/location/http.ts

```diff
@@ -0,0 +1,86 @@
+import {
+  SnapManifest,
+  assertIsSnapManifest,
+  VFile,
+  HttpSnapIdStruct,
+} from '@metamask/snaps-utils';
+import { assert, assertStruct } from '@metamask/utils';
+
+import { ensureRelative } from '../../utils';
+import { SnapLocation } from './location';
+
+export default class HttpLocation implements SnapLocation {
+  // We keep contents separate because then we can use only one Blob in cache,
+  // which we convert to Uint8Array when actually returning the file.
+  //
+  // That avoids deepCloning file contents.
+  // I imagine ArrayBuffers are copy-on-write optimized, meaning
+  // in most often case we'll only have one file contents in common case.
+  private readonly cache = new Map<string, { file: VFile; contents: Blob }>();
+
+  private validatedManifest?: VFile<SnapManifest>;
+
+  private readonly url: URL;
+
+  constructor(url: URL) {
+    assertStruct(url.toString(), HttpSnapIdStruct, 'Invalid Snap Id: ');
+    this.url = url;
+  }
+
+  async manifest(): Promise<VFile<SnapManifest>> {
+    if (this.validatedManifest) {
+      return this.validatedManifest.clone();
```

Comment:
Because it's a cache, and we want to allow all other parts to modify the vfile as they see fit, for example adding more metadata to `vfile.data`, in that case the original cache shouldn't be modified.